### DISCUSSION TITLE
Does Jinja read the file from the hard drive every time it is requested?

### DISCUSSION THREAD

**Discussion by boludoz** [pallets/jinja#2042] [General]
Does Jinja read the file from the hard drive every time it is requested? Is there a cache made in C++ with unordered maps or something like that?

**Comment by ThiefMaster** [comment]
Why would one use C++ for that? Python `dict` and similar structures are already implemented in C and highly efficient.

Anyway, as per [the FAQ](https://jinja.palletsprojects.com/en/3.1.x/faq/):

> Jinja also makes extensive use of caching. Templates are cached by name after loading, so future uses of the template avoid loading. The template loading itself uses a bytecode cache to avoid repeated compiling. The caches can be external to persist across restarts. Templates can also be precompiled and loaded as fast Python imports.

**Reply by boludoz** [reply]
Yes, but it caught my attention that every time I modify an HTML it is read again.

**Reply by boludoz** [reply]
I don't have to implement any cache or anything like that then?

**Reply by boludoz** [reply]
You're right, dict is similar to unordered map, I hadn't thought of that.

**Reply by ernstki** [reply]
> Yes, but it caught my attention that every time I modify an HTML it is read again.

Is that not the desired effect?
